Tactics/SymContext.lean

```diff
@@ -0,0 +1,230 @@
+        | throwError "Unknown local variable `{state}`"
+      pure (Expr.fvar decl.fvarId)
+    | none => mkFreshExprMVar (Expr.const ``ArmState [])
+
+  -- Find `h_run` and infer `runSteps` from it
+  let sf ← mkFreshExprMVar none
+  let runSteps ← mkFreshExprMVar (Expr.const ``Nat [])
+  let h_run_type ← mkEq sf (←mkAppM ``run #[runSteps, stateExpr])
+  let h_run ← findLocalDeclUsernameOfTypeOrError h_run_type
+
+  -- Unwrap and reflect `runSteps`
+  let some runSteps := (← instantiateMVars runSteps).nat?
+    | throwError "Expected a numeric literal, found:\n\t{runSteps}\nIn\n\t {h_run} : {h_run_type}"
+  -- TODO: we should allow all ground terms here, not just literals.
+  -- For example, we sometimes use `sf = run someProgram.length s0`
+
+  -- At this point, `stateExpr` should have been assigned (if it was an mvar),
+  -- so we can unwrap it to get the underlying name
+  let stateExpr ← instantiateMVars stateExpr
+  let state ← state?.getDM <| do
+    let .fvar state := stateExpr
+      | throwError
+  "Expected a free variable, found:
+    {stateExpr}
+  We inferred this as the initial state because we found:
+    {h_run} : {← instantiateMVars h_run_type}
+  in the local context.
+
+  If this is wrong, please explicitly provide the right initial state, as `sym {runSteps} at ?s0`
+  "
+    let some state := lctx.find? state
+      /- I don't expect this error to be possible in a well-formed state, but you never know -/
+      | throwError "Failed to find local variable for state {stateExpr}"
+    pure state.userName
+
+  -- Try to find `h_program`, and infer `program` from it
+  let program ← mkFreshExprMVar none
+  let h_program_type ← mkEq (← mkAppM ``ArmState.program #[stateExpr]) program
+  let h_program ← findLocalDeclUsernameOfTypeOrError h_program_type
+
+  -- Assert that `program` is a(n application of a) constant, and find its name
+  let program := (← instantiateMVars program).getAppFn
+  let .const program _ := program
+    | throwError "Expected a constant, found:\n\t{program}\nIn: {h_program} : {h_program_type}"
+  /-
+    TODO: assert that the expected `stepi` theorems have been generated for `program`
+  -/
+
+  -- Then, try to find `h_pc`
+  let pc ← mkFreshExprMVar (← mkAppM ``BitVec #[toExpr 64])
+  let h_pc_type ← mkEq (← mkAppM ``r #[(.const ``StateField.PC []), stateExpr]) pc
+  let h_pc ← findLocalDeclUsernameOfTypeOrError h_pc_type
+
+  -- Unwrap and reflect `pc`
+  let pc ← instantiateMVars pc
+  let pc ←
+    try
+      reflectBitVecLiteral 64 pc
+    catch e =>
+      throwError "{e.toMessageData}\nIn\n\t{h_pc} : {h_pc_type}"
+
+  return inferStatePrefixAndNumber {
+    state, h_run, runSteps, program, h_program, pc, h_pc,
+    h_err := none,
+    h_sp := none,
+  }
+where
+  findLocalDeclUsernameOfType? (expectedType : Expr) : MetaM (Option Name) := do
+    let some h_run ← findLocalDeclWithType? expectedType
+      | return none
+    let decl := (← getLCtx).get! h_run
```

Comment:
```suggestion
    let decl := (← getLCtx).get! h_run
```

the `get!` is suspicious, a comment explaining why this is OK would be nice.

Reply:
Added a comment